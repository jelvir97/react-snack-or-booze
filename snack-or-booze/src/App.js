import React, { useState, useEffect } from "react";
import { Route, Routes, } from "react-router-dom";
import "./App.css";
import Home from "./Home";
import SnackOrBoozeApi from "./Api";
import NavBar from "./NavBar";
import Menu from "./Menu";
import MenuItem from "./MenuItem";
import AddItemForm from "./AddItemForm";
import NotFound from "./NotFound";

function App() {
  // if true, loading page is rendered
  const [isLoading, setIsLoading] = useState(true);
  // items contains arrays of snacks and drinks.
  const [items, setItems] = useState({});

  /**
   * Gets snacks and drinks from db on first render.
   */
  useEffect(() => {
    async function getSnacks() {
      let items = await SnackOrBoozeApi.getItems();
      setItems(items)
      setIsLoading(false);
    }
    getSnacks();
  }, []);

  /**
   * Sends post request to SnackOrBoozeApi
   * Updates items to include the newly added snack or drink
   */
  const addItem = async(formData)=>{
    const {data} = await SnackOrBoozeApi.addItem(formData)
    const newItem = {name:data.name, description:data.description, recipe:data.recipe, serve:data.serve}
    setItems( ()=>{
        return {
            ...items,
            [formData.resource]: [...items[formData.resource], newItem]
        }
    })
  }

  /**
   * If true, loading page is rendered
   */
  if (isLoading) {
    return <p>Loading &hellip;</p>;
  }

  return (
    <div className="App">
      
        <NavBar />
        <main>
          <Routes>
            <Route exact path="/" 
              element={<Home items={items} />}
            />
            
            {/** Snacks Routes */}
            <Route exact path="/snacks"
              element={<Menu items={items.snacks} resource="snacks" title="Food" />}
            />
            <Route path="/snacks/:id"
              element={<MenuItem items={items.snacks} cantFind="/snacks" />}
            />

            {/** Drinks Routes */}
            <Route exact path="/drinks"
              element={<Menu items={items.drinks} resource="drinks" title="Drinks" />}
            />
            <Route path="/drinks/:id"
              element={<MenuItem items={items.drinks} cantFind="/drinks" />}
            />

            <Route exact path="/add"
              element={<AddItemForm addItem={addItem} items={items}/>}
            />

            {/** Not Found */}
            <Route path="*"
              element={<NotFound />}
            />
          </Routes>
        </main>
      
    </div>
  );
}

export default App;
