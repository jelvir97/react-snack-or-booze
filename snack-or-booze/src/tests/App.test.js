import { render, screen, waitFor, act} from "@testing-library/react";
import  "@testing-library/jest-dom"
import { MemoryRouter } from 'react-router-dom';
import App from "../App"
import userEvent from "@testing-library/user-event";

jest.mock('../Api', () => ({
    ...jest.requireActual('../Api'),
    getItems: ()=>({
        snacks: [
            {
                id:'test-snack-1',
                name: "test snack 1",
                description:"test snack description 1",
                recipe: "test snack recipe 1",
                serve: "test snack serve 1"
            },
            {
                id:'test-snack-2',
                name: "test snack 2",
                description:"test snack description 2",
                recipe: "test snack recipe 2",
                serve: "test snack serve 2"
            }
        ],
    
        drinks: [
            {
                id:'test-drink-1',
                name: "test drink 1",
                description:"test drink description 1",
                recipe: "test drink recipe 1",
                serve: "test drink serve 1"
            },
            {
                id:'test-drink-2',
                name: "test drink 2",
                description:"test drink description 2",
                recipe: "test drink recipe 2",
                serve: "test drink serve 2"
            }
        ]
    }),
    addItem: (formData)=>({
        data: { ...formData, id: "test-id" }
    })
}));

describe("smoke and snapshot tests", ()=>{
    
    it("should render", async ()=>{
        await waitFor(()=>{
            render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            )
        })

    })

    it("should match snapshot", async ()=>{
        await waitFor(()=>{
            const {asFragment} = render(
                <MemoryRouter initialEntries={['/']}>
                    <App />
                </MemoryRouter>
            )

            expect(asFragment).toMatchSnapshot()
        })

    })

})

describe('routes working',()=>{
    it("should render home page", async ()=>{
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={["/"]}>
                    <App />
                </MemoryRouter>
            )       
        })
        expect(await screen.findByText(`Welcome to Silicon Valley's premier dive cafe!`)).toBeInTheDocument()
    })

    it("should render add item form page", async ()=>{
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={["/add"]}>
                    <App />
                </MemoryRouter>
            )       
        })
        expect(await screen.findByText(`Add Menu Item`)).toBeInTheDocument()
    })

    it("should render food menu page", async ()=>{
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={["/snacks"]}>
                    <App />
                </MemoryRouter>
            )       
        })
        expect(await screen.findByText(`Food Menu`)).toBeInTheDocument()
        expect(await screen.findByText(`test snack 1`)).toBeInTheDocument()
    })

    it("should render drink menu page", async ()=>{
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={["/drinks"]}>
                    <App />
                </MemoryRouter>
            )       
        })
        expect(await screen.findByText(`Drinks Menu`)).toBeInTheDocument()
        expect(await screen.findByText(`test drink 1`)).toBeInTheDocument()
    })

    it("should render not-found page", async ()=>{
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={["/invalid-url"]}>
                    <App />
                </MemoryRouter>
            )       
        })
        expect(await screen.findByText(`Hmm... what your looking for doesn't exist.`)).toBeInTheDocument()
    })
})

describe('testing navigation', ()=>{
    it("should navigate to menus", async()=>{
        const user = userEvent.setup()
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={['/']}>
                    <App />
                </MemoryRouter>
            )
        })

        const snackMenuBtns = await screen.findAllByText('Snacks')
        user.click(snackMenuBtns[0])
        expect(await screen.findByText('Food Menu'))
    })

    it('should navigate to MenuItem', async()=>{
        const user = userEvent.setup()
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={['/']}>
                    <App />
                </MemoryRouter>
            )
        })

        const drinkMenuBtns = await screen.findAllByText('Drinks')
        user.click(drinkMenuBtns[0])
        expect(await screen.findByText('Drinks Menu'))

        const testDrinkLink = await screen.findByText('test drink 1')
        user.click(testDrinkLink)
        expect( await screen.findByText('Recipe:'))
        expect( await screen.findByText('test drink 1'))

    })

    it('should navigate to AddItemForm', async()=>{
        const user = userEvent.setup()
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={['/']}>
                    <App />
                </MemoryRouter>
            )
        })

        const addFormLink = await screen.findByText('Add')
        user.click(addFormLink)
        expect(await screen.findByText('Add Menu Item'))
    })

    it('should navigate to Home', async()=>{
        const user = userEvent.setup()
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={['/invalid-url']}>
                    <App />
                </MemoryRouter>
            )
        })

        const goHomeBtn = await screen.findByText('Go Home')
        user.click(goHomeBtn)
        expect(await screen.findByText(`Welcome to Silicon Valley's premier dive cafe!`))
    })
})

describe('test adding new menu items',()=>{
    it('should add new snack', async()=>{
        const user = userEvent.setup()
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={["/add"]}>
                    <App />
                </MemoryRouter>
            )
        })
        //const type = await screen.findByLabelText('Type') * don't need to change *
        const description = await screen.findByLabelText('Description')
        const recipe = await screen.findByLabelText('Recipe')
        const serve = await screen.findByLabelText('Serve')
        const name = await screen.findByLabelText('Name')
        const submitBtn = await screen.findByText('Submit')

        await act(async()=>{
            await user.type(description, 'test snack 3')
            await user.type(recipe, 'test snack 3')
            await user.type(serve, 'test snack 3')
            await user.type(name, 'test snack 3')

            await user.click(submitBtn)
        })

        expect( await screen.findByText('Food Menu')).toBeInTheDocument()
        expect( await screen.findByText('test snack 3')).toBeInTheDocument()
    })

    it("should fail because of conflicting item name", async()=>{
        //Item with the name ${formData.name} already exists.
        const user = userEvent.setup()
        await waitFor(()=>{
            render(
                <MemoryRouter initialEntries={["/add"]}>
                    <App />
                </MemoryRouter>
            )
        })

        const description = await screen.findByLabelText('Description')
        const recipe = await screen.findByLabelText('Recipe')
        const serve = await screen.findByLabelText('Serve')
        const name = await screen.findByLabelText('Name')
        const submitBtn = await screen.findByText('Submit')

        await act(async()=>{
            await user.type(description, 'test snack 2')
            await user.type(recipe, 'test snack 2')
            await user.type(serve, 'test snack 2')
            await user.type(name, 'test snack 2')

            await user.click(submitBtn)
        })

        expect(await screen.findByText(`Add Menu Item`)).toBeInTheDocument()
        expect(await screen.findByText(`Item with the name test snack 2 already exists.`)).toBeInTheDocument()
    })
})