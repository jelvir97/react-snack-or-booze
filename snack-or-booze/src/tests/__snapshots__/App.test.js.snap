// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests should match snapshot 1`] = `[Function]`;
